kernel/src/engine/default/storage.rs

```
@@ -1,38 +1,141 @@
-#[cfg(feature = "cloud")]
-use hdfs_native_object_store::HdfsObjectStore;
 use object_store::parse_url_opts as parse_url_opts_object_store;
 use object_store::path::Path;
 use object_store::{Error, ObjectStore};
 use url::Url;
 
+use std::collections::HashMap;
+use std::sync::{Arc, OnceLock, RwLock};
+
+/// Alias for convenience
+type ClosureReturn = Result<(Box<dyn ObjectStore>, Path), Error>;
+/// This type alias makes it easier to reference the handler closure(s)
+///
+/// It uses a HashMap<String, String> which _must_ be converted in our [parse_url_opts] because we
+/// cannot use generics in this scenario.
+type HandlerClosure = Arc<dyn Fn(&Url, HashMap<String, String>) -> ClosureReturn + Send + Sync>;
+/// hashmap containing scheme => handler fn mappings to allow consumers of delta-kernel-rs provide
+/// their own url opts parsers for different scemes
+type Handlers = HashMap<String, HandlerClosure>;
+/// The URL_REGISTRY contains the custom URL scheme handlers that will parse URL options
+static URL_REGISTRY: OnceLock<RwLock<Handlers>> = OnceLock::new();
+
+/// Return to fully initialied URL handlers to allow extensible provisioning of [object_store] URL
+/// handlers
+fn get_url_handlers<'life>() -> &'life RwLock<Handlers> {
+    URL_REGISTRY.get_or_init(|| RwLock::new(HashMap::default()))
+}
+
+/// Insert a new URL handler for [parse_url_opts] with the given `scheme`. This allows users to
+/// provide their own custom URL handler to plug new [object_store::ObjectStore] instances into
+/// delta-kernel
+pub fn insert_url_handler(
+    scheme: impl AsRef<str>,
+    handler_closure: HandlerClosure,
+) -> Result<(), Error> {
+    // Make sure the map is initialized before trying to get a mutable reference to it
+    let registry = get_url_handlers();
+    if let Ok(mut lock) = registry.write() {
+        (*lock).insert(scheme.as_ref().into(), handler_closure);
+    } else {
+        panic!("Failed to acquire lock for adding a URL handler!");
```

can we return an error instead of `panic!`?

---

I am kind of indifferent here, I think failing to insert a URL handler is a pretty fatal error, nothing will work properly if you were expecting to insert a URL handler and that failed

---

Agreed, but given that panics are really a pain, especially over ffi, we aim for a "no panic" policy if possible in kernel, and should rather return an error. The engine gets to decide how fatal it is :)

---

Fatal for Delta queries that need the URL handler perhaps, but surely not the engine as a whole? A panic would bring down the whole engine.